README.md

```diff
@@ -29,7 +29,11 @@ To put the rM2 in recovery, you will need:
 ## Next steps
 * to recover the ssh password, mount the home partition `grep Developer <mountpoint>/root/.config/remarkable/xochitl.conf`
 * to find a mountpoint run `mount | grep home`
-* TODO not tested, but something like `dd if=raw of=/dev/mountedroot1`
+* your `dmesg` output will show the letter assigned to the block device (`sdc` will be used for the following examples)
+* Make full backup of device `sudo dd if=/dev/sdc of=rm2.raw bs=4096 status=progress`
+* Open the `uboot.env` file from the root of the `BOOT` partition in a text editor. You'll see something like `active_partition=3`. This indicates that your remarkable boots from sdc3, so we'll use sdc2 to restore the `sdc3` partition. 
+* Backup both system partitions using `sudo dd if=/dev/sdc2 of=rm2-sdc2.raw bs=4096 status=progress` and `sudo dd if=/dev/sdc3 of=rm2-sdc3.raw bs=4096 status=progress`
+* Restore the `sdc3` partition with the backup of `sdc2` using `sudo dd if=rm2-sdc2.raw of=/dev/sdc3 bs=4096 status=progress` assuming your active partition is `3`. If your active partition is `2` restore it using `sudo dd if=rm2-sdc3.raw of=/dev/sdc2 bs=4096 status=progress`
```

Comment:
This should probably also outline how to restore from an update image. You can use codexctl to download the image, and then extract the image with https://github.com/ddvk/remarkable-update/tree/main/extractor or programatically with https://github.com/Eeems-Org/remarkable-update-fuse

Reply:
Thanks @nr23730 and @Eeems -- this was really helpful in downgrading from the latest 3.11 beta -- I'd like to note that the only image extraction method that _actually worked_ was using https://github.com/ddvk/remarkable-update/tree/main/extractor, https://github.com/ddvk/remarkable-update/blob/main/switch.sh, and this dd command:

```
dd if=rawimage.ext4 bs=4096 of=/dev/mmcblk2p3 && ./switch.sh && reboot
```
Any other method of extracting the image resulted in the tablet just rebooting back to the beta firmware partition.

To clarify, I was able to do this via SSH, didn't need the HW adapter 😄

Reply:
Since you are doing this over SSH, it's not really relevant to what the instructions here are aimed at.

> I'd like to note that the only image extraction method that _actually worked_ was using https://github.com/ddvk/remarkable-update/tree/main/extractor, https://github.com/ddvk/remarkable-update/blob/main/switch.sh, and this dd command:
> 
> ```
> dd if=rawimage.ext4 bs=4096 of=/dev/mmcblk2p3 && ./switch.sh && reboot
> ```
> 
> Any other method of extracting the image resulted in the tablet just rebooting back to the beta firmware partition.

What other method did you attempt to use to extract the image? I only listed two options, one of which requires you to write your own python script to handle the extraction, which I know works, as it's part of the [automated testing in the repo](https://github.com/Eeems-Org/remarkable-update-fuse/blob/a0ad943b160b305165a9975e311a7dc2e603dc4d/test.py#L210-L222).


---

Side-note. 3.11+ has a new update mechanism, which changes the image format completely. New tools and instructions will need to be provided for how to get a raw image from the new update images.

Reply:
> Since you are doing this over SSH, it's not really relevant to what the instructions here are aimed at.

Yeah, I just felt the need to jot this down _somewhere_ in case others on 3.11+ ran into the same issues. Spent a while trying to understand what was going wrong...

> What other method did you attempt to use to extract the image?

I originally tried `codextl extract` via https://github.com/Jayy001/codexctl, [which uses](https://github.com/Jayy001/codexctl/blob/main/codexctl/__init__.py#L560-L569) your remarkable fuse package.

After failing a bunch with that image, my last thought after a few dozen tries the other day was that there must be a difference between versions/logic/etc in the three methods (codextl, extractor.py, diy py with fuse package) and tried directly using your package at the python prompt and also using the extractor.py script. The image from extractor.py was the only one that actually worked.

I did see one reddit post of a user on a much lower version (2.x) who was unable to get their downgrade to work, until it randomly did after a few tries -- it's possible this was just an intermittent failure (or that my images were damaged when `scp`ing to the RM2)? 

It seemed like it _really_ didn't want to downgrade from 3.11 and kept changing the firmware vals back...

Also, originally I hadn't been using the block size param with `dd`, but I went back and tried again after seeing this post (which was really helpful to confirm I was taking the right steps). I did try again with my original extracted image, but block size alone wasn't enough to fix it.

> Side-note. 3.11+ has a new update mechanism

I did see that, but figured I could still flash the other partition and downgrade. I _was_ worried something with the boot process had changed, but I think it was just _really_ stubborn.

---

Very thankful for y'all in the community keeping these tools alive -- I've got toltec/KOReader/etc running now now problem😄

Reply:
> > What other method did you attempt to use to extract the image?
> 
> I originally tried `codextl extract` via https://github.com/Jayy001/codexctl, [which uses](https://github.com/Jayy001/codexctl/blob/main/codexctl/__init__.py#L560-L569) your remarkable fuse package.
> 
> After failing a bunch with that image, my last thought after a few dozen tries the other day was that there must be a difference between versions/logic/etc in the three methods (codextl, extractor.py, diy py with fuse package) and tried directly using your package at the python prompt and also using the extractor.py script. The image from extractor.py was the only one that actually worked.

I don't see an open issue on either repository for the errors you were getting. So Jayy and I have no way of fixing whatever the issue you encountered was.

The logic extractor.py and my stuff uses is the same, my stuff just adds extra validation checks and performance optimizations for random access.

> 
> I did see one reddit post of a user on a much lower version (2.x) who was unable to get their downgrade to work, until it randomly did after a few tries -- it's possible this was just an intermittent failure (or that my images were damaged when `scp`ing to the RM2)?
> 
> It seemed like it _really_ didn't want to downgrade from 3.11 and kept changing the firmware vals back...

The extracted images would need to be applied with `dd`, not the update mechansim, so I'm not sure how it would not work if you got an image to extract. Did you actually have errors with extracting the image, or were you still trying to use the old update mechansim to apply an update file, and used the extracted images instead?

> 
> Also, originally I hadn't been using the block size param with `dd`, but I went back and tried again after seeing this post (which was really helpful to confirm I was taking the right steps). I did try again with my original extracted image, but block size alone wasn't enough to fix it.

`dd` doesn't need to have a block size to work. It's just used for optimizing the speed, or calculating the size when doing a partial copy or from an infinite source like `/dev/zero`.

Reply:
> I don't see an open issue on either repository for the errors you were getting. So Jayy and I have no way of fixing whatever the issue you encountered was.

There were no errors -- each method produced an image file without issue, but only the last attempt (with the extractor.py) actually worked.

> Did you actually have errors with extracting the image, or were you still trying to use the old update mechansim to apply an update file, and used the extracted images instead?

Neither -- with the new update mechanism in 3.11, I didn't bother with the OTA methods and tried `dd` straight away.

The only real ideas I have at this point are either:
- earlier images were corrupted when transferring to the RM2, or,
- the u-boot firmware env vars were being reset somehow, intermittently

As far as I could tell, everything was working as expected -- image extraction, writing with `dd`, and the `fw_setenv` calls all looked good -- but the RM2 would reboot and fallback to the 3.11 partition every time.

Maybe 3.11 is "stickier", or maybe my images were corrupted the first few times... 🤷‍♂️  Regardless, don't need to waste any more time on this, appreciate y'all's efforts on all these tools.

When it finally _did_ work, I just wanted to post it somewhere in the hopes of potentially avoiding https://xkcd.com/979/...

Reply:
I would still appreciate an issue opened with more information, like the sha256sum of the update file, and of the produced image. We can't fix the issue if people don't report them with enough information to work with.

So far, I've used both mechanisms to produce non-corrupted images, and I have an automated test for rmufuse to make sure that it can extract images as expected.

I appreciate trying to avoid xkcd 979, but I do think that only giving partial information here and not reporting the issue to the broken upstream things you tried using that _should work_ isn't really helping anything. Especially when you are just posting on a pull request comment, so this will be lost as time goes on as well.

Reply:
I spent a couple of hours and sorted out what the issue was: https://github.com/Eeems-Org/remarkable-update-fuse/issues/7

Reply:
@Eeems Went to start grabbing checksums and saw you had fixed this. Apologies for not providing those earlier, in hindsight that wasn't too helpful... Really appreciate you taking the time to dig in and thank you for fixing the issue!

Is there somewhere I can buy you a beer or a coffee (donate)? Thanks again!

Reply:
https://teamtrees.org/ or https://teamseas.org/ could always use more donations. I don't personally take any money for things I do as a hobby, it would change the relationship I have with it to be something that I'm obligated to do, instead of something that I do for fun.

Reply:
![L1ZY1LN2L](https://github.com/ddvk/remarkable2-recovery/assets/1571956/cb8501a5-aff9-4022-8504-636d517b95bb)